petastorm/pytorch.py

```diff
@@ -423,3 +425,160 @@ def __enter__(self):
+        returned to a user by the ``BatchedDataLoader`` until the queue is full. After that, batches of `batch_size`
+        will be created by uniformly sampling the shuffling queue. Once no more samples are available from the data
+        reader, the shuffling queue is allowed to be consumed till no further samples are available.
+
+        Note that the last returned batch could have less then ``batch_size`` samples.
+
+        NOTE: if you are using ``make_batch_reader``, this shuffling queue will be randomizing the order of the
+        entire batches and not changing the order of elements within a batch. This is likely not what you intend to do.
+
+        This class does not support special types that are not supported in PyTorch (decimal/string).
+
+        :param reader: petastorm Reader instance
+        :param batch_size: the number of items to return per batch; factored into the len() of this reader
+        :param transform_fn: an optional callable to convert batches from the reader to PyTorch tensors
+        :param shuffling_queue_capacity: Queue capacity is passed to the underlying :class:`tf.RandomShuffleQueue`
+          instance. If set to 0, no shuffling will be done.
+        :param async_batch_queue_capacity: Number of batches to be produced in advance. If async_batch_queue_capacity
+            == 0, shuffle buffer will generate batches based on request. If async_batch_queue_capacity > 0, shuffle
+            work is done by an asynchronous thread to generate batches in advance. It is suggested to set this value
+            to be non-zero if shuffling is bottleneck for dataloader.
+        :param inmemory_cache_all: This is an boolean that indicates whether the data should be
+            cached in memory or not. When this cache is enabled, expect smaller than requested
+            batch size on the epoch boundaries. Enabling this cache also requres that shuffling_queue_capacity to be
+            large enough to hold all data.
+        :param num_epochs: Number of epochs to load when in memory cache is enabled. If
+            set to None, loader will loads batches indefinitely.
+        :param async_batch_queue_capacity: Capacity of queue to store shuffled batches. Shuffle work is done by an
+            asynchronous thread to produce batches into a queue.
+        """
+        super(AsyncBatchedDataLoader, self).__init__(reader, batch_size=batch_size, transform_fn=transform_fn,
+                                                     shuffling_queue_capacity=shuffling_queue_capacity,
+                                                     inmemory_cache_all=inmemory_cache_all,
+                                                     num_epochs=num_epochs)
+        self._async_batch_queue = Queue(async_batch_queue_capacity)
+        self._finished_event = Event()
+        self._shuffle_thread = None
+        self._started = False
+
+    def _worker(self):
+        for batch in self._iter_impl_worker():
```

Comment:
What happens when `self._iter_impl_worker` raises an exception? Could it be that `batch = self._async_batch_queue.get()` will be stuck forever, or there is some mechanism that prevents that?

Let's cover the exception raising in a unittest? (could put some raise in a transform func to make it easy to test)